File: 2024/day1/day1.mjs
import fs from "node:fs";

const sample1 = fs.readFileSync("sample1.txt", { encoding: "utf-8" });

console.log(part1(sample1));

/**
 *
 * @param {String} fileContent
 */
function part1(fileContent) {
  const lines = fileContent.trim().split("\r\n");
  const leftDists = [];
  const rightDists = [];

  lines.map((l) => {
    const [left, right] = l.split("   ");
    leftDists.push(left);
    rightDists.push(right);
  });

  leftDists.sort((a, b) => a - b);
  rightDists.sort((a, b) => a - b);

  let distSum = 0;

  for (let i = 0; i < leftDists.length; i++) {
    const leftDist = leftDists[i];
    const rightDist = rightDists[i];

    distSum += Math.abs(leftDist - rightDist);
  }

  return distSum;
}

File: 2024/day1/day1.rs
use itertools::Itertools;

fn get_distances(input: String) -> (Vec<usize>, Vec<usize>) {
    input
        .trim()
        .lines()
        .map(|line| {
            let (left, right) = line.split_once("   ").unwrap();
            (
                left.parse::<usize>().unwrap(),
                right.parse::<usize>().unwrap(),
            )
        })
        .collect()
}

fn part1(input_data: String) -> usize {
    let (mut left_distances, mut right_distances) = get_distances(input_data);

    left_distances.sort_unstable();
    right_distances.sort_unstable();

    left_distances
        .iter()
        .zip(right_distances)
        .map(|(l, r)| l.abs_diff(r))
        .sum()
}

fn part2(input_data: String) -> usize {
    let (left_distances, right_distances) = get_distances(input_data);

    let frequencies = right_distances.into_iter().counts();

    left_distances
        .iter()
        .filter_map(|n| frequencies.get(n).map(|f| f * n))
        .sum()
}

#[cfg(test)]
mod tests {
    use utils::read_file;

    use super::*;

    #[test]
    fn test1() {
        assert_eq!(part1(read_file("sample1.txt")), 11);
    }

    #[test]
    fn test2() {
        assert_eq!(part2(read_file("sample1.txt")), 31);
    }
}

fn main() {
    utils::run(1, &["sample1.txt", "input.txt"], &part1, &part2);
}
